sql/telemetry_derived/clients_histogram_aggregates_v1/query.sql

```diff
@@ -193,5 +148,9 @@ SELECT
   app_version,
   app_build_id,
   channel,
-  udf_merged_user_data(old_aggs, new_aggs) AS histogram_aggregates
+  new_aggs IS NOT NULL AS updated_on_last_run,
+  CASE
+    WHEN new_aggs IS NULL THEN udf_clear_histogram_aggregates(old_aggs)
+    ELSE udf_merged_user_data(old_aggs, new_aggs)
```

Comment:
Yes that's correct. This is to avoid keeping data unnecessarily. I will add a comment.

Reply:
Does this mean that historical aggregates are cleared when a client doesn't exhibit a histogram for a session?

Example:

order|new_aggs|old_aggs
-|-|-
0|[1,0,1]|null
1|[0,2,1]|[1,0,1]
2|null|[1,2,1]
3|[1,0,1]|null

In (2), `old_aggs` are the sum of the `new_aggs` in (0) and (1). But because `new_aggs` is null, the `old_aggs` in (3) is null. It seems like the correct behavior should be to have `old_aggs=[1,2,1]` in (3).

Reply:
I think part of the confusion might be that there are `old_aggs` and `new_aggs` which are different from `old_aggregates` and `new_aggregates` and I need to do some variable name adjustment. Here is my attempt at describing things given what is there now:

For every client+metric+key+process etc dimension combination, a client stores in clients_histogram_aggregates both an `old_aggregates` histogram and a `new_aggregates` histogram. At any given time, all values for `new_aggregates` should represent the most recent histogram for this client+metric dimension combination. Only if a client+metric dimension combination changed on the last run should it have a value for `old_aggregates` which represents its value before the most recent update.

When new client data is pulled in and a full outer join is done on old and new client data, there are 3 possible scenarios.
1. The dimension combination had no new data but it existed previously so `new_aggs` in the join is NULL but old_aggs is not.
2. The dimension combination is entirely new so `old_aggs` in the join is NULL but `new_aggs` is not.
3. The dimension combination has new data and there was previously existing data for it so both `old_aggs` and `new_aggs` are not null.
Given these possibilities, `updated_on_last_run` is set to true if `new_aggs` is not null. In this case, we also need to merge `old_aggs` and `new_aggs`. So we need the following result in the merge (this is done by udf_merged_user_data()):
1. We need to call map_sum() on the `new_aggregates` of both the `new_aggs` and the `old_aggs` to compute the newest histograms - these will be labelled as the `new_aggregates`
2. We need to save the `new_aggregates` of `old_aggs` and label as the `old_aggregates`
If `new_aggs` is null, then these rows need to be cleared of old data because no updates will be made using them in `histogram_bucket_counts`.

Let me know if that answers your question.

Reply:
> If new_aggs is null, then these rows need to be cleared of old data because no updates will be made using them in histogram_bucket_counts.

Okay, kind of see what's going on here, but it's difficult to understand the behavior in all of the edge cases. Before, only the `aggregate` tables were stateful, but now `bucket_counts` tables rely on an incremental computation.

To rephrase question: are there no longer running sums of client histograms? It seems like the histogram_aggregates table no longer keeps histogram histories per client, only aggregated diffs between computations of the histogram buckets. 

If this is true, it is no longer possible to recover bucket counts to a complete state if it needs to be recomputed. I think this is problematic. Bucket counts would also need to be persisted with backups for disaster recovery in order to maintain "correct" state about buckets.

Reply:
There are still running sums of client histograms. 

Looking at the previous code for `clients_histogram_aggregates` for `udf_merged_user_data()`, we see that `old_aggs` and `new_aggs` both had an `aggregates` array.  The array `aggregates` has effectively been renamed `new_aggregates` and all of the data should behave exactly the same in `new_aggregates` as it did in `aggregates` - incrementally creating a running sum.

`old_aggregates` is now added so that it keeps the last version of the running sum before the most recent update.

So it should actually be possible to use nearly exactly the same query for `clients_histogram_bucket_counts` applied to `new_aggregates` instead of `aggregates` and get the same results (I've tested this briefly myself and it seems to look correct.)

I do agree that bucket counts should have a 3-day archive just like histogram aggregates now that it's incremental.